### ISSUE TITLE
Small-Omega Update in Synaptic Intelligence (SI)

### ISSUE THREAD

**Issue by graldij** [aimagelab/mammoth#50]
Ciao!

First of all, thank you so much for this awesome repo.

Secondly, I have possibly noticed a flaw in the update rule of small_omega in the SI implementation. Specifically, small_omega is incremented solely based on the regularized gradients, instead of both the regularized and non-regularized multiplied ([L61 of si.py](https://github.com/aimagelab/mammoth/blob/f42d05377bc473c173ed4a710ec2f628d3bc8a6f/models/si.py#L61)):
```
        self.small_omega += self.args.lr * self.net.get_grads().data ** 2
```
As a reference, you can see the official implementation of Zenke at [this](https://github.com/ganguli-lab/pathint/blob/86c5e07c5603d6a44c2f53585c19ee70773ef15c/pathint/protocols.py#L34) line.

If I was not clear, or if you have any questions, please let me know!

Best,
Jacopo

**Comment by loribonna** [comment]
Hey @graldij ,
you are probably right. In our implementation we made a couple of assumptions on `\theta'(t)` of Eq. 3 of [the main paper ](https://arxiv.org/pdf/1703.04200), where the parameter update would be equal to the gradient times the learning rate, thus yielding `self.args.lr * self.net.get_grads().data ** 2`. This seems also to follow Eq. 10 and Sec. C of the main paper, but correct me if I'm wrong.

Regardless, I agree that the correct implementation should use the unregularized gradients, since in the paper `L(\theta)` is only the classification loss.

We will update the code shortly to fix this.

**Comment by graldij** [comment]
This is how I interpret it: $g_k$ requires the unregularized gradient as it is a proxy for the contribution to that task's (unregularized) loss, while $\delta_k$ is the actual change in parameters, which is a consequence of the actual optimization problem with regularization. Therefore both regularized and unregularized gradients should be considered in the update of small_omega. 

Let me know if this makes sense to you

**Comment by graldij** [comment]
I see what you mean. Correct me if I am wrong, but this does not imply that "the way" we get to those final parameters is the same (nor that we will actually ever get to those final parameters, e.g. due to early stopping), right?

**Comment by loribonna** [comment]
I'm not sure because the small_omega takes into account all the path between the start and the end of the task. 
Also from further inspection of the claim 

> if learning were to be performed on the surrogate loss instead of the actual loss function, it would result in the same final parameters and change in loss during training.

It does not seem that exact (check Fig. 3 of the paper). 

At this point I think the best way is to follow the implementation of Zenke and compute the gradient on the unregularized loss and the real displacement, as you suggested.

The main issue here is that I think the paper may be open to interpretation: the small_omega of Eq. 3 is defined on the unregularized loss, so from a formal standpoint the displacement should follow that loss, making it far more similar to the empirical Fisher.

**Comment by loribonna** [comment]
@graldij this got automatically closed but let me know if you think it still has issues and I will re-open it.
Also, thank you for your help!

**Comment by graldij** [comment]
No worries!

I'm afraid I don't agree with the current fix. In particular, it does not seem that the actual step is done with the gradients of the regularized loss. Also, I don't get what line 63 does, maybe a loss.backward() is missing in `get_penalty_grads`?

**Comment by loribonna** [comment]
I'll re-open it to make it easier to find.
The current implementation computes directly the closed form of the gradients, so instead of needing to call the backward twice (first on the unregularized loss and then on the total one) it sums the gradient of the regularization term.
This is possible since the regularization does not depend on the input data and is quite easy to derive.

**Comment by graldij** [comment]
My bad, I did not notice that `get_penalty_grads()` already contains the gradient, in contrast to `penalty()` (which is now unused and can be deleted). 

Thanks for the responsiveness!

**Comment by loribonna** [comment]
No problem! We will keep the `penalty()` function to serve as a reference.
Thank you for your help with Mammoth!
